src/bin/julialauncher.rs

```diff
@@ -300,6 +302,181 @@ fn get_override_channel(
+        // We want to find the first non-switch argument or the first argument after `--`
+        let mut skip_next = false;
+        for (i, arg) in args.iter().skip(1).enumerate() {
+            if skip_next {
+                skip_next = false;
+            } else if arg == "--" {
+                program_file = args.get(i + 1).cloned();
+                break;
+            } else if arg.starts_with("--") {
+                if !no_arg_long_switches.contains(&arg.as_str()) && !arg.contains('=') {
+                    skip_next = true;
+                }
+            } else if arg.starts_with("-") {
+                let arg: Vec<char> = arg.chars().skip(1).collect();
+                if arg.iter().all(|&c| no_arg_short_switches.contains(&c)) {
+                    continue;
+                }
+                for (j, &c) in arg.iter().enumerate() {
+                    if no_arg_short_switches.contains(&c) {
+                        continue;
+                    } else if j < arg.len() - 1 {
+                        break;
+                    } else {
+                        // `j == arg.len() - 1`
+                        skip_next = true;
+                    }
+                }
+            } else {
+                program_file = Some(arg.clone());
+                break;
+            }
+        }
+        if let Some(program_file) = program_file {
+            let mut path = PathBuf::from(program_file);
+            path.pop();
+            while !path.join("Project.toml").exists() && !path.join("JuliaProject.toml").exists() {
+                if !path.pop() {
+                    return None;
+                }
+            }
+            return Some(path);
+        } else {
+            return None;
+        }
+    } else if project.starts_with('@') {
+        let depot = match std::env::var("JULIA_DEPOT_PATH") {
+            Ok(val) => match val.split(':').next() {
+                Some(p) => PathBuf::from(p),
+                None => dirs::home_dir().unwrap().join(".julia"),
+            },
+            _ => dirs::home_dir().unwrap().join(".julia"),
+        };
+        let path = depot.join("environments").join(&project[1..]);
+        if path.exists() {
+            return Some(path);
+        } else {
+            return None;
+        }
+    } else {
+        return Some(PathBuf::from(project));
+    }
+}
+
+fn julia_version_from_manifest(path: PathBuf) -> Option<Version> {
+    let manifest = if path.join("Manifest.toml").exists() {
+        path.join("Manifest.toml")
+    } else if path.join("JuliaManifest.toml").exists() {
+        path.join("JuliaManifest.toml")
+    } else {
+        return None;
+    };
+    let file = std::fs::File::open(manifest).ok()?;
+    let reader = std::io::BufReader::new(file);
+    // This is a somewhat bootleg way to parse the manifest,
+    // but since we know the format it should be fine.
+    for line in reader.lines() {
```

Comment:
We should use a proper TOML parser here, I'm sure there is one for Rust.

Reply:
We can, I just don't think it's needed. Since the manifest has a rather prescribed format, a manifest that doesn't work with this will cause issues with Julia anyway.

Reply:
It just is very, very hacky, and I would much prefer to do the right thing here and use a tested library. https://docs.rs/toml/0.8.19/toml/ exists and seems easy enough.

Reply:
I don't know about all those "very"s :stuck_out_tongue:, with how nice and simple TOML is. There are exactly two ways this could produce an incorrect result:
1. A manifest with no initial `julia_version` key, and then a `julia_version` table key that's non-indented. However, (1) Pkg doesn't use arbitrary keys, and (2) Julia's TOML writer always indents nested table keys.
2. A multi-line string that contains a line starting with `julia_version = value`.

In other words, the only manifests that this will not work with are manifests that _aren't Julia manifests_, and we can't produce correct results then anyway.

Reply:
Eh, I can't be bothered to discuss this any further. The `toml` crate is now used.